### PR TITLE
[NXP MCUX] Driver for 802.15.4 over FSCI protocol

### PR DESCRIPTION
**Pull request by bogdan-davidoaia**
The driver is meant to run on hexiwear_k64. To have it working you need to run NXP's MAC FSCI app on hexiwear_kw40.

You can get the MAC FSCI app by installing the [Kinetis KW40Z Connectivity Software](https://www.nxp.com/webapp/Download?colCode=KW40Z-CONNECTIVITY-SOFTWARE&appType=license&location=null&fpsp=1&WT_TYPE=Protocol%20Stacks&WT_VENDOR=FREESCALE&WT_FILE_FORMAT=exe&WT_ASSET=Downloads&fileExt=.exe&Parent_nodeId=1432854896956716810497&Parent_pageType=product) and flashing _ConnSw\tools\binaries\MAC_FSCI_app_frdmkw40z.bin_.

Update by @pfalcon:

From https://www.nxp.com/docs/en/reference-manual/CONNFWKRM.pdf , section 3.10, p.89:

> The Freescale Serial Communication Interface (FSCI) is [...] a protocol that allows monitoring and extensive testing of the protocol layers interfaces. It also allows separation of the protocol stack between two protocol layers in a two processing entities setup: the host processor (typically running the upper layers of a protocol stack) and the Black Box application (typically containing the lower layers of the stack, serving as a modem).

### CODE REVIEW

**Comment by bogdan-davidoaia** [comment]
Still need to do some more testing, but uploading it now for more people to have access to it.

**Comment by bogdan-davidoaia** [comment]
Did a rebase and added config files for echo server/client. Some debugging is still required as the echo client/server doesn't work for me yet (tested against a mcr20a).

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
This driver could be used on other platforms, not just hexiwear, so it should be renamed to ieee802154_fsci or something similar. It probably shouldn't be ieee802154_fsci_uart since it looks like you could add spi to this driver rather than creating a new spi-based fsci driver. The calls to uart api's are pretty well-contained.

**Comment by bogdan-davidoaia** [comment]
Updated naming and fixed a few issues. Echo server becomes stuck after doing RX; will do more debugging tomorrow.

File: dts/arm/hexiwear_k64.dts [hunk 1]
```diff
@@ -33,6 +33,9 @@
 #ifdef CONFIG_BLUETOOTH
 		zephyr,bt-uart = &uart4;
 #endif
+#ifdef CONFIG_IEEE802154_FSCI
+		zephry,uart-pipe = &uart4;
```
Comment (dleach02) [inline dts/arm/hexiwear_k64.dts]: misspelled zephyr in line 37 which causes the build to fail.

Comment (bogdan-davidoaia) [inline dts/arm/hexiwear_k64.dts]: Fixed.

**Comment by bogdan-davidoaia** [comment]
Although not sure what the root cause is, the issue with the echo client/server seems to be related to 6low fragmentation. If I force the UDP payload to be 61 bytes or less, the sample works fine.

File: samples/net/ieee802154/hw/prj_fsci.conf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+CONFIG_NET_STATISTICS_PERIODIC_OUTPUT=n
+
+CONFIG_NET_L2_IEEE802154=y
+CONFIG_NET_DEBUG_L2_IEEE802154=y
+CONFIG_NET_L2_IEEE802154_FRAGMENT=y
+#CONFIG_NET_DEBUG_L2_IEEE802154_FRAGMENT=y
+CONFIG_NET_L2_IEEE802154_SECURITY=n
+CONFIG_NET_L2_IEEE802154_SHELL=y
+
+CONFIG_IEEE802154_FSCI=y
```
Comment (jukkar) [inline samples/net/ieee802154/hw/prj_fsci.conf]: Note that PR #957 makes some changes how the prj files are used. So you might need to tweak these prj files a bit.

**Comment by jukkar** [comment]
> Although not sure what the root cause is, the issue with the echo client/server seems to be related to 6low fragmentation. If I force the UDP payload to be 61 bytes or less, the sample works fine.

Is this issue in RX or TX path?

**Comment by bogdan-davidoaia** [comment]
I didn't get to dig that deep into it and now don't have access to the hardware anymore. Someone from NXP/Linaro will continue to work on it.

What I can say is that when using a packet sniffer, the server doesn't send any packets in response to packets that have the 6LoWPAN fragmentation header.

Can someone check if the echo client/server sample with 6LowPAN fragmentation works when using NRF?

File: drivers/ieee802154/Kconfig [hunk 3]
```diff
@@ -56,10 +56,10 @@ config SYS_LOG_IEEE802154_DRIVER_LEVEL
 
 source "drivers/ieee802154/Kconfig.cc2520"
 
-source "drivers/ieee802154/Kconfig.kw41z"
-
```
Comment (tbursztyka) [inline drivers/ieee802154/Kconfig]: isn't this making Kconfig.kw41z an orphaned file?

File: drivers/ieee802154/ieee802154_fsci.c [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+	buf = frag->data;
+	fs = (struct ieee802154_fcf_seq *)buf;
+	memset(fs, 0, sizeof(struct ieee802154_fcf_seq));
+	buf += sizeof(struct ieee802154_fcf_seq);
+
+	fs->fc.frame_type = IEEE802154_FRAME_TYPE_DATA;
+	fs->fc.frame_version = IEEE802154_VERSION_802154_2006;
+	fs->fc.pan_id_comp = 1;
+	fs->sequence = data->dsn;
+
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_fsci.c]: why do you need to create the fcf/seq header again? That's included in all 15.4 frame, it's never build on reception, just parsed (and that's up to the 15.4 MAC layer to do so, not the driver).

File: drivers/ieee802154/ieee802154_fsci.c [hunk 5]
```diff
@@ -0,0 +1,705 @@
...
+		fsci_context_data.rx_state = FSCI_RX_STATE_SYNC;
+		break;
+	}
+
+	*off = 0;
+	return buf;
+}
+
+static void fsci_fill_crc(clientPacket_t *pkt)
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_fsci.c]: the hardware is not able to fill this in by itself?
Even the cheapest 15.4 crappy chip usually does.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Hum, is this FSMI a full-MAC driver framework?

Fact that you rebuilt the FCF/seq on rx, and remove it away on tx, seems to indicate that. So I stopped my review after the 2nd patch.

Then you'll need to revise this completely: do not hook up a 15.4 full-mac driver to the 15.4 soft-mac (l2 layer), it just piles up redundant logic here. Instead create a new full-mac dedicated 15.4 l2 layer that would play nicely with your driver. With or without a dedicated API in between, up to you. Such l2 driver will handle 6lo, unless your framework is able to do it by itself as well (does fcsi embeds 6lo?)

File: drivers/ieee802154/ieee802154_fsci.c (hunk 5 above)
Comment (dleach02) [inline drivers/ieee802154/ieee802154_fsci.c]: This is a checksum that is specific to the FSCI protocol and not 15.4. The protocol is fairly simple with a start deliminator, header with length, payload and checksum.

**Comment by pfalcon** [comment]
Following the Linaro Connect, continuing to wrap my head around this. So, we do have "native" (accessing hardware directly) 15.4 driver for kw41z, which was also extended to kw40z: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ieee802154/ieee802154_kw41z.c (written by @bogdan-davidoaia , just as this patch). Perhaps it might still have some issues, but it's native driver at least. But it's intended to run when Zephyr runs natively on an MCU.

hexiwear_k64 has a bit different architecture: it has two MCUs: an application one and "network processor". That's not a unique case for Zephyr, e.g. we have 96b_carbon, which also has an app MCU and BLE controller offloaded to another MCU.

So, the driver in this patch assumes that hexiwear_k64's kw40z MCU will run a 15.4 binary blob stack, and controls it using FSCI protocol (which in this context is similar to Bluetooth HCI protocol). But that's of course not the only way to do it. For example, on 96b_carbon we run Zephyr on both MCUs, and they communicate using HCI over SPI.

We could do the same for hexiwear_k64 (which of course will require more effort, but will progress Zephyr further, and will allow to control the devices "better"). But even if we do, we'll still need to communicate somehow between app MCU and network MCU. And unless there's a "more standard" protocol to do that for 15.4, reusing FSCI makes good sense (unless there can be issues with reusing it on other vendors' MCUs). Actually, reusing FSCI instead of inventing own protocol would be better, as it will allow to switch between the binary block stack and Zephyr stack, and thus improve interoperability.

All in all, this driver seems to be worth to continue to work on.

**Comment by pfalcon** [comment]
@dleach02, @MaureenHelm, @galak : Please see if the above comment summarizes the state of affairs well. If it does, I'd probably move it to the PR description (I already added reference to the FSCI doc there).

**Comment by tbursztyka** [comment]
> We could do the same for hexiwear_k64 (which of course will require more effort, but will progress Zephyr further, and will allow to control the devices "better"). But even if we do, we'll still need to communicate somehow between app MCU and network MCU. And unless there's a "more standard" protocol to do that for 15.4, reusing FSCI makes good sense (unless there can be issues with reusing it on other vendors' MCUs). Actually, reusing FSCI instead of inventing own protocol would be better, as it will allow to switch between the binary block stack and Zephyr stack, and thus improve interoperability.

That would be nice to have a common way to deal with 15.4 MCU, the same way as BT does. Unfortunately, unlike BT with HCI, FSCI is not a standard at all, and under NXP's IP if I am not mistaken. So it'll be nearly impossible to push it to other.

Instead, it would require that a group of people like 15.4 manufacturers, under IEEE watch for instance, gather and decide about a standard for it. Imo taking HCI as a base would be just perfect, more over if you consider the fact some MCU can do both BT/15.4 - on 2.4GHz at least - sharing the same radio core, like NRF's one actually. There had been some talks about doing something like that ~1-2 years ago but nothing that went further.

> All in all, this driver seems to be worth to continue to work on.

Nobody said the contrary :)

But, looks like the current status is going a little bit off. If FSCI is a way to deal with NXP's 15.4 MCU, it will mean the MCU owns the 15.4 MAC (a "Full MAC"). Thus: the host FCSI driver **must not** hook itself up to Zephyr's 802.15.4 L2 which is also a 15.4 MAC implementation (a "Soft-MAC"). It would unnecessarily bloat the host with code that is anyway running already on the MCU. Instead, you'll need an FSCI L2 layer (or, if it can be generalized properly, why not "IEEE802154_FULL_MAC" L2). Such L2 would need however to implement the same net mgmt part as the Soft-MAC L2 does (that's a detail though, it's quite straight forward).
Unless FSCI allows accessing bare metal 15.4 chip too. Then 2 drivers are required: one like the current proposal, and a second with the full mac l2 thingy I talked about.

Now, as you talk about the reverse way: running Zephyr on the MCU as well, providing the same FSCI interface. It would require a different driver that exposes Zephyr 15.4 Soft-MAC through FSCI interface. Also, I believe it would require NXP to agree on this, unless they already give the details on how to make your own firmware for the MCU.

**Comment by dleach02** [comment]
Here are the issues (as I understand them); as pointed out, the KW40Z Black Box blob is a full 802.15.4 MAC. The FSCI protocol happens to be the current method NXP uses to communicate to the connectivity frameworks we produce. The KW40Z FSCI driver is translating the 802.15.4 data frames into 802.3 data frames (problem) to be sent by the KW40Z blob. The first problem then is that the driver is at the bottom of the Zephyr 802.15.4 stack interfacing to the top of the Kinetis KW40Z stack. The reason Bogdan went down this path was because of the limited support of the KW40Z in Zephyr. I suspect he was thinking he could just talk to the NXP firmware in the KW40Z via FSCI but having an 802.15.4 stack on top of another 802.15.4 stack is not going to work out well.

There are some options:

1) Have the FSCI driver interface at a different, and higher level of the 802.15.4 stack. Basically, interface at the top edge of the management layer as a FullMAC driver interface to 802.15.4 and directly hook up the 6LoWPAN/IP stack. I tried to probe how useful this really would be. I don't have a good sense of what is out there in the market in the 802.15.4 space and if the only use of this interface is this Hexiwear K64/KW40 system then it probably isn't worth the effort to do.

2) Create new KW40Z firmware that exposes the lower edge of the 802.15.4 stack to the host system. This presents a logistical problem of where do we store this NXP firmware image.

I don't have a good picture of what kind of support we actually have for the KW40Z in Zephyr. I was initially told that it is very limited and some of the problems relate back to NXP's lack of support for the platform as well. I suggested that we drop this ticket because there is going to be a new platform coming out that includes the K64F and the KW41Z. The KW41Z has better support in both Zephyr and within NXP. It would also afford the Zephyr project the opportunity to define and implement a standard IPC mechanism to facilitate communication between the chips. I do not recommend using FSCI as the communication method.

**Comment by pfalcon** [comment]
@tbursztyka : Thanks for the comments. I probably should have mentioned why I'm popping on this ticket. Previously, I didn't follow 15.4 work closely (it was just a hardware enablement task for NXP engineers), but now it was confirmed that Linaro is interested in 15.4 in general, and was given a hardware to work on it (kw41z). So, I'm trying to wrap my head around what was done previously, in particular, why we need this driver, if we already head https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ieee802154/ieee802154_kw41z.c .

>> All in all, this driver seems to be worth to continue to work on.
> Nobody said the contrary :)

Well, actually, at the last Linaro Connect (and Zephyr mini-minisummit how Anas called it), I heard @MaureenHelm and @dleach02 discussed possibility of just closing this (that's what I heard, I probably wasn't part of the whole discussion).

> FSCI is not a standard at all, and under NXP's IP if I am not mistaken. So it'll be nearly impossible to push it to other.

I may imagine that would be the case. Well, it's up to the NXP folks to consider it.

> Instead, it would require that a group of people like 15.4 manufacturers, under IEEE watch for instance

Well, that didn't happen before, so unlikely happen soon ;-). So, the choices left is either to reuse another such protocol (e.g. Linux was solving this problem long before us), or make an adhoc one. The latter always works of course.

> If FSCI is a way to deal with NXP's 15.4 MCU, it will mean the MCU owns the 15.4 MAC (a "Full MAC"). Thus: the host FCSI driver must not hook itself up to Zephyr's 802.15.4 L2 which is also a 15.4 MAC implementation (a "Soft-MAC").

So far, I didn't get that far. I just looked up that FSCI is a trivial, generic wire protocol (header, operation, length, data, checksum). I didn't look what's being transferred there in the case of kw41z 15.4 binary blob firmware. Even if's "FullMAC" implementation, I would (naively) think that it still might offer operations to work with PHY level, but if it doesn't, oops.

**Comment by pfalcon** [comment]
> I don't have a good picture of what kind of support we actually have for the KW40Z in Zephyr.

Well, there're commits like e392ef2285f9b80858a6c53c37a73f8e0a681339.

> I don't have a good sense of what is out there in the market in the 802.15.4 space and if the only use of this interface is this Hexiwear K64/KW40 system then it probably isn't worth the effort to do.

I can only second this - unless there's a specific requirement from NXP to support "15.4 offload", we'd better base on and concentrate around https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ieee802154/ieee802154_kw41z.c, and architecture when Zephyr runs on both chips of multi-chip boards like hexiwear.

> I suggested that we drop this ticket [...]. It would also afford the Zephyr project the opportunity to define and implement a standard IPC mechanism to facilitate communication between the chips. I do not recommend using FSCI as the communication method.

Sounds good to me.

**Comment by tbursztyka** [comment]
Ok, up to you guys.

**Comment by galak** [comment]
@MaureenHelm @dleach02 should we close this out?

**Comment by nashif** [comment]
outdated, closing. Reopen with updates/rebase if needed.

**Pull request closed without merging**